dali/kernels/common/simd.h

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_COMMON_SIMD_H_
+#define DALI_KERNELS_COMMON_SIMD_H_
+
+#ifdef __SSE2__
+#include <emmintrin.h>
+#endif
+
+#include <cstddef>
+#include <type_traits>
+#include "dali/core/force_inline.h"
+
+namespace dali {
+namespace kernels {
+namespace simd {
+
+#ifdef __SSE2__
+
+template <int n>
+struct float4x {
+  __m128 v[n];  // NOLINT
+};
+
+template <int n>
+struct i128x {
+  __m128i v[n];  // NOLINT
+};
+
+using float4x1 = float4x<1>;
+using float4x2 = float4x<2>;
+using float4x4 = float4x<4>;
+using i128x1 = i128x<1>;
+using i128x2 = i128x<2>;
+using i128x4 = i128x<4>;
+
+inline __m128i clamp_round(__m128 f, float lo, float hi) {
+  f = _mm_min_ps(_mm_max_ps(f, _mm_set1_ps(lo)), _mm_set1_ps(hi));
+  return _mm_cvtps_epi32(f);  // round
+}
+
+inline void store(int8_t *i8, i128x4 iv) {
+  __m128i sv0 = _mm_packs_epi32(iv.v[0], iv.v[1]);
+  __m128i sv1 = _mm_packs_epi32(iv.v[2], iv.v[3]);
+  _mm_storeu_si128(reinterpret_cast<__m128i*>(i8), _mm_packs_epi16(sv0, sv1));
+}
+
+inline void store(uint8_t *u8, i128x4 iv) {
+  __m128i sv0 = _mm_packs_epi32(iv.v[0], iv.v[1]);
+  __m128i sv1 = _mm_packs_epi32(iv.v[2], iv.v[3]);
+  _mm_storeu_si128(reinterpret_cast<__m128i*>(u8), _mm_packus_epi16(sv0, sv1));
+}
+
+inline void store(int16_t *i16, i128x2 iv) {
+  _mm_storeu_si128(reinterpret_cast<__m128i*>(i16), _mm_packs_epi32(iv.v[0], iv.v[1]));
+}
+
+inline void store(uint16_t *u16, i128x2 iv) {
+  __m128i sh = _mm_set_epi32(0, 0, 0, 1);
```

Comment:
There's no instruction that can pack 32-bit values to 16-bit without _signed_ saturation - neither is there a byte shuffle.

Reply:
Can you add this as a comment to this function?

Reply:
So this function just cuts the MSB?

Reply:
Per-element, it does this:
a = (unsigned)x >> 1
b = x & 1
c = saturate<int16>(a)
out = c << 1 | b